### PR TITLE
concourse terraform

### PR DESCRIPTION
**Pull request by prabalsharma**
Integrates terraform to existing qa pipeline
Updates upgrade url to 1.4.1

### CODE REVIEW

File: qa-pipelines/qa-pipeline.yml.erb [hunk 1]
```diff
@@ -227,26 +255,38 @@ jobs:
...
     # block, so that when tests fail we have a chance of examining why things are
     # failing.
-    <% if enable_cf_teardown %>
+    <% if enable_cf_teardown && !enable_terraform %>
     - task: cf-teardown
       file: ci/qa-pipelines/tasks/cf-teardown.yml
       timeout: 1h
       params: *common-task-params
-    <% end %> # enable_cf_teardown
+    <% end %> # enable_cf_teardown and not enable_terraform
```
Comment (HartS) [inline qa-pipelines/qa-pipeline.yml.erb]: `and` and `&&` are not operationally equivalent in ruby, so let's stick with `&&` to reflect the condition being closed

File: cap-terraform/.gitignore [hunk 2]
```diff
@@ -0,0 +1,8 @@
+**/creds/*
+.terraform
+*.backup
+*.tfstate
+scf*/
```
Comment (HartS) [inline cap-terraform/.gitignore]: Where are files matching this path prefix created?

File: cf-ci-orchestration/Dockerfile [hunk 3]
```diff
@@ -24,6 +24,10 @@ RUN true \
     git-core \
     bash-completion docker-bash-completion systemd-bash-completion \
     python-xml python-pyOpenSSL python-devel \
+    && curl -LO "https://releases.hashicorp.com/terraform/0.12.2/terraform_0.12.2_linux_amd64.zip" \
```
Comment (HartS) [inline cf-ci-orchestration/Dockerfile]: When possible, changes to the Dockerfile should happen at the end in a new run step to leverage caching, and avoid rebuilding layers (faster pushes to docker hub and faster pulls from docker hub)

File: qa-pipelines/flags.yml [hunk 4]
```diff
@@ -4,11 +4,14 @@
 # The settings in this file will also be valid for our most complete pipeline
 # (which performs an upgrade and all tests)
 
+# Use terrafrom to deploy AKS cluster
```
Comment (HartS) [inline qa-pipelines/flags.yml]: terrafrom

Comment (HartS) [inline qa-pipelines/flags.yml]: terrafrom

File: qa-pipelines/qa-pipeline.yml.erb [hunk 5]
```diff
@@ -91,6 +91,15 @@ aliases:
     KUBE_ORGANIZATION: ((organization))
     MAGIC_DNS_SERVICE: ((magic-dns-service))
 
+    # Terraform AKS Configuration
+    ARM_CLIENT_ID: ((arm-client-id))
```
Comment (HartS) [inline qa-pipelines/qa-pipeline.yml.erb]: If these vars are only needed by the terraform task, they shouldn't go in the common task params

File: qa-pipelines/qa-pipeline.yml.erb [hunk 6]
```diff
@@ -135,8 +152,20 @@ jobs:
...
+      <% if enable_terraform %>
+      try:
+        do:
+          - task: delete-aks-resource-group
+            file: ci/qa-pipelines/tasks/delete-aks-resource-group.yml
+            params:
+              << : *common-task-params
+          - put: pool.kube-hosts
+            params: {remove: pool.kube-hosts}
+      <% end %> # enable_terraform
```
Comment (HartS) [inline qa-pipelines/qa-pipeline.yml.erb]: Use an else here instead

File: qa-pipelines/qa-pipeline.yml.erb [hunk 7]
```diff
@@ -135,8 +152,20 @@ jobs:
...
+      <% if !enable_terraform %>
       put: pool.kube-hosts
       params: {release: pool.kube-hosts}
+      <% end %> # not enable_terraform
```
Comment (HartS) [inline qa-pipelines/qa-pipeline.yml.erb]: `# enable_terraform`, not `# not enable_terraform`

File: qa-pipelines/qa-pipeline.yml.erb [hunk 8]
```diff
@@ -227,26 +255,38 @@ jobs:
...
     - task: cf-teardown
       file: ci/qa-pipelines/tasks/cf-teardown.yml
       timeout: 1h
       params: *common-task-params
-    <% end %> # enable_cf_teardown
+    <% end %> # enable_cf_teardown and not enable_terraform
+    <% if !enable_terraform %>
     - put: pool.kube-hosts
       params: {release: pool.kube-hosts}
+    <% end %> # not enable_terraform
```
Comment (HartS) [inline qa-pipelines/qa-pipeline.yml.erb]: Same as above, use `# !enable_terraform`

File: qa-pipelines/qa-pipeline.yml.erb [hunk 9]
```diff
@@ -227,26 +255,38 @@ jobs:
...
     <% end %> # status_reporting
     <% if status_reporting %>
     on_failure:
       put: status.src
       params:
-        <<: *status-params-<%= avail %>
+        << : *status-params-<%= avail %>
         state: failure
     <% end %> # status_reporting
+    <% if enable_terraform %>
```
Comment (HartS) [inline qa-pipelines/qa-pipeline.yml.erb]: Shouldn't terraform teardown task templating be informed by the `enable-teardown` flag also?

File: qa-pipelines/tasks/delete-aks-resource-group.sh [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
```
Comment (HartS) [inline qa-pipelines/tasks/delete-aks-resource-group.sh]: Why can't we just use terraform for teardown?

File: qa-pipelines/tasks/delete-aks-resource-group.yml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+inputs:
+- name: ci
+- name: pool.kube-hosts
+
+params:
+  KUBE_REGISTRY_HOSTNAME: ""
+  KUBE_REGISTRY_USERNAME: ""
+  KUBE_REGISTRY_PASSWORD: ""
+  KUBE_ORGANIZATION: ""
+  HA: false # true, false, or "scaled"
```
Comment (HartS) [inline qa-pipelines/tasks/delete-aks-resource-group.yml]: annotations aren't needed here. Also, is this param even being used by anything in the delete-terraform task?

File: cap-terraform/.gitignore (hunk 2 above)
Comment (prabalsharma) [reply cap-terraform/.gitignore]: thats from parent directory: https://github.com/SUSE/cap-terraform/blob/master/.gitignore
I won't touch that as Satadru experimented with that. If that gets cleaned in cap-terraform I will remove it from here

File: cf-ci-orchestration/Dockerfile (hunk 3 above)
Comment (prabalsharma) [reply cf-ci-orchestration/Dockerfile]: Do you want a change on this?

File: qa-pipelines/flags.yml (hunk 4 above)
Comment (prabalsharma) [reply qa-pipelines/flags.yml]: oops

Comment (prabalsharma) [reply qa-pipelines/flags.yml]: lol
thanks

File: qa-pipelines/qa-pipeline.yml.erb (hunk 5 above)
Comment (prabalsharma) [reply qa-pipelines/qa-pipeline.yml.erb]: I think we should keep just one prams for all tasks. It keeps the qa-pipeline.erb clean. I was planning to put every thing in common params after refactor2 merge.

File: qa-pipelines/qa-pipeline.yml.erb (hunk 7 above)
Comment (prabalsharma) [reply qa-pipelines/qa-pipeline.yml.erb]: no its  <% if !enable_terraform %>

File: qa-pipelines/qa-pipeline.yml.erb (hunk 6 above)
Comment (prabalsharma) [reply qa-pipelines/qa-pipeline.yml.erb]: I kept it this way : 

1.  we can just add additional condition in addition to !enable_terraform when required.
2. keep it consistent with linehttps://github.com/SUSE/cf-ci/blob/d78b41d1862cb40665707215fc3bf0d4cbc30145/qa-pipelines/qa-pipeline.yml.erb#L165
3. easy to read compared to else. At this point, I prefer readability in qa-pipeline.erb we desperately need it :)

File: qa-pipelines/qa-pipeline.yml.erb (hunk 7 above)
Comment (prabalsharma) [reply qa-pipelines/qa-pipeline.yml.erb]: anh u meant # !enable_terraform

File: qa-pipelines/qa-pipeline.yml.erb (hunk 9 above)
Comment (prabalsharma) [reply qa-pipelines/qa-pipeline.yml.erb]: no. The terraform CI work follow this guideline:  "QA CI don't care if user want to enable-teardown or not. CI will tear it down."

File: qa-pipelines/tasks/delete-aks-resource-group.sh (hunk 10 above)
Comment (prabalsharma) [reply qa-pipelines/tasks/delete-aks-resource-group.sh]: Cause that will require us to use public cloud to save the state of terraform. Which we don't want to do. An architectural decision was made to not use such resource as its an over kill for our needs. 
In addition, terrform  teardown might leave leftover resources as seen in past whereas use az to delete resource group is very effective.

File: qa-pipelines/tasks/delete-aks-resource-group.yml (hunk 11 above)
Comment (prabalsharma) [reply qa-pipelines/tasks/delete-aks-resource-group.yml]: I just copied this from cf-teardown.yml. As the requirement of such annotations were triggered when using new terrform which I am not using.
Please let me know which line to delete here.

File: cf-ci-orchestration/Dockerfile (hunk 3 above)
Comment (HartS) [reply cf-ci-orchestration/Dockerfile]: No it's fine, let's just keep it in mindfor future changes here

File: qa-pipelines/qa-pipeline.yml.erb (hunk 5 above)
Comment (HartS) [reply qa-pipelines/qa-pipeline.yml.erb]: This is only needed for one terraform task, so it should go in that task parameter list. It's not a common param, and it puts unnecessary things in the pipeline definition (after templating).

File: qa-pipelines/flags.yml (hunk 4 above)
Comment (HartS) [reply qa-pipelines/flags.yml]: Change requested

Comment (HartS) [reply qa-pipelines/flags.yml]: change requested

File: qa-pipelines/qa-pipeline.yml.erb (hunk 5 above)
Comment (HartS) [reply qa-pipelines/qa-pipeline.yml.erb]: Change requested

File: qa-pipelines/qa-pipeline.yml.erb (hunk 1 above)
Comment (HartS) [reply qa-pipelines/qa-pipeline.yml.erb]: Change requested

File: qa-pipelines/qa-pipeline.yml.erb (hunk 7 above)
Comment (HartS) [reply qa-pipelines/qa-pipeline.yml.erb]: change requested

**Pull request merged**